11ty/eleventy
# Include template data file and directory data file dependencies in Eleventy Serverless bundles

Resolves #2222 

This PR introduces a new `eleventy-app-localdata-modules.js` to the serverless bundle, which includes imports from template data files and directory data files.

I've introduced a `writeBundlerLocalDependenciesFile` helper function specifically for local data files' modules because I noticed that the `"eleventy.dataFiles"` event (where `writeBundlerLocalDependenciesFile` is being called) is called multiple times, presumably once per template. `writeFileSync`, as used in `writeBundlerDependenciesFile`, overwrites the file if it exists — so continuing to use that for template/directory data files means that only the last local data file's dependencies would be included in the bundle. Rather than extend the existing `writeBundlerDependenciesFile` abstraction, I opted to create a special `writeBundlerLocalDependenciesFile` helper just for this particular case.

Please let me know if there's anything I can do to improve this!

## Validation Steps

1. Clone [this repro repository](https://github.com/BenDMyers/serverless-dependency-repro) referenced in #2222
2. Delete `netlify/functions` directory
3. Install this version of Eleventy
4. `npm run dev`
5. Verify the following:
- Project builds successfully
- Serverless function's `eleventy-bundler-modules.js` file contains the statement `require("./eleventy-app-localdata-modules.js");`
- Serverless function has an `eleventy-app-localdata-modules.js` file.
- `eleventy-app-localdata-modules.js` requires `is-odd` and `uuid`.
6. Add more dependencies to template data files and directory data files.
7. Verify these dependencies are added to `eleventy-app-localdata-modules.js`.
